### PR TITLE
board: emsdp: add a new arc based board support

### PR DESCRIPTION
**Pull request by IRISZZW**
add the support of emsdp board, a new arc based board.

The DesignWare® ARC® EM Software Development Platform (SDP) is a flexible platform
for rapid software development on ARC EM processor-based subsystems. It is intended
to accelerate software development and debug of ARC EM processors and subsystems for
a wide range of ultra-low power embedded applications such as IoT, sensor fusion,
and voice applications. 

The EMSDP includes an FPGA-based hardware board with commonly used peripherals and 
interfaces for extensibility. Downloadable platform packages containing different hardware configurations enable the board to be programmed with different ARC EM processors and 
subsystems. Now it has support for ARC EM4, EM5D, EM6, EM7D, EM9D and EM11D core 
configuration.

The details about the EMSDP board: [EMSDP](https://www.synopsys.com/dw/ipdir.php?ds=arc-em-software-development-platform)

This PR includes the following support:

- processor core support

- uart-driver

- gpio-driver

### CODE REVIEW

File: boards/arc/emsdp/board.dtsi [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			label = "DIP SW1 - Switch 3";
+		};
+		switch3: switch_3 {
+			/* gpio flags need validation */
+			gpios = <&gpio0 3 GPIO_INT_ACTIVE_LOW>;
+			label = "DIP SW1 - Switch 4";
+		};
+	};
+
+};
```
Comment (dbkinder) [inline boards/arc/emsdp/board.dtsi]: It's good to end files with a newline

File: boards/arc/emsdp/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+This command still uses openocd and gdb to load the application elf file to EM
+Software Development Platform, but it will load the application and immediately run.
+If power is removed, the application will be lost since it wasn't written to flash.
+
+Most of the time you will not be flashing your program but will instead debug
+it using openocd and gdb. The program can be download via the USB cable into
+the code and data memories.
+
+Release Notes
```
Comment (dbkinder) [inline boards/arc/emsdp/doc/index.rst]: There's a heading here with no content?

File: boards/arc/emsdp/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+*************
+
+References
```
Comment (dbkinder) [inline boards/arc/emsdp/doc/index.rst]: Same here (a heading with no content).   If you want the links below to actually show up in the output here, add a ``target-notes`` directive before the first link:
```
References
**********
.. target-notes::

.. _Digilent Pmod Modules: http://store.digilentinc.com/pmod-modules
.. _Putty website: http://www.putty.org
```

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arc/emsdp/board.dtsi (hunk 1 above)
Comment (IRISZZW) [reply boards/arc/emsdp/board.dtsi]: already done

File: boards/arc/emsdp/doc/index.rst (hunk 2 above)
Comment (IRISZZW) [reply boards/arc/emsdp/doc/index.rst]: removed

File: boards/arc/emsdp/doc/index.rst (hunk 3 above)
Comment (IRISZZW) [reply boards/arc/emsdp/doc/index.rst]: added

**Review by vonhust — CHANGES_REQUESTED** [review]
pls squash some commits to have a better commit history

**Review by vonhust — APPROVED** [review]
Sanitycheck tests are ok.  It's ok for me

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks!

**Review by nashif — CHANGES_REQUESTED** [review]
Please squash the last commit into where the changes have originated.

**Comment by IRISZZW** [comment]
> Please squash the last commit into where the changes have originated.

already done  and rebased with the master branch

File: boards/arc/emsdp/Kconfig.board [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+#
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+config BOARD_EMSDP
+	bool "EM Software Development Platform"
+	depends on SOC_ARC_EMSDP
+	help
+	  The ARC EM Software Development Platform (emsdp) is an FPGA based development platform intended to support ARC licenses in developing their software for the ARC EM processor family and ARC EM Subsystems. It has the support for ARC EM4, EM5D, EM6, EM7D, EM9D and EM11D processors. ARC EM Enhanced Security Package (ESP) and ARC EM Subsystems (DFSS, SCSS, DSS) are also supported.
```
Comment (SebastianBoe) [inline boards/arc/emsdp/Kconfig.board]: The help text should have newlines, see other multi-line help texts for how to do indentation correctly.

File: soc/arc/snps_emsdp/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+elseif(CONFIG_SOC_EMSDP_EM7D)
+zephyr_compile_options(-mmpy-option=6)
+elseif(CONFIG_SOC_EMSDP_EM7D_ESP)
+zephyr_compile_options(-mmpy-option=6)
+elseif(CONFIG_SOC_EMSDP_EM9D)
+zephyr_compile_options(-mmpy-option=6)
+zephyr_compile_options_ifdef(CONFIG_FLOAT -mfpu=fpus_all)
+elseif(CONFIG_SOC_EMSDP_EM11D)
+zephyr_compile_options(-mmpy-option=6)
+zephyr_compile_options_ifdef(CONFIG_FLOAT -mfpu=fpuda_all)
```
Comment (SebastianBoe) [inline soc/arc/snps_emsdp/CMakeLists.txt]: indentation

File: boards/arc/emsdp/Kconfig.board (hunk 4 above)
Comment (IRISZZW) [reply boards/arc/emsdp/Kconfig.board]: already done.

File: soc/arc/snps_emsdp/CMakeLists.txt (hunk 5 above)
Comment (IRISZZW) [reply soc/arc/snps_emsdp/CMakeLists.txt]: already done.

**Review by nashif — APPROVED** [review]

File: boards/arc/emsdp/emsdp_em4.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+identifier: emsdp
```
Comment (ruuddw) [inline boards/arc/emsdp/emsdp_em4.yaml]: All the emsdp boards have the same identifier (emsdp), but these should be unique.
Please change into emsdp_em4 here, and _em6 etc. in the other board yaml files.

**Review by ruuddw — COMMENTED** [review]
In principle ok, but one thing should still be fixed: unique identifiers for the board yaml files.

File: boards/arc/emsdp/emsdp_em4.yaml (hunk 6 above)
Comment (IRISZZW) [reply boards/arc/emsdp/emsdp_em4.yaml]: already done.

**Comment by vonhust** [comment]
ping for merge

**Comment by vonhust** [comment]
@ruudw, PR of EMSDP is ready,  pls review, approve and merge

**Review by ruuddw — APPROVED** [review]

**Comment by andrewboie** [comment]
needs a rebase, otherwise this LGTM

**Comment by IRISZZW** [comment]
> needs a rebase, otherwise this LGTM

done

**Comment by andrewboie** [comment]
Error: Aborting due to non-whitelisted Kconfig warning 'warning: default on the choice symbol
CPU_ARCEM (defined at boards/arc/emsdp/Kconfig.defconfig:13, arch/arc/Kconfig:19) will have no
effect, as defaults do not affect choice symbols'. Note: If this warning doesn't point to an actual
problem, you can add it to the whitelist at the top of
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/kconfig/kconfig.py.

File: boards/arc/emsdp/emsdp_em4.dts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+&uart0 {
+	status = "okay";
+	current-speed = <115200>;
+};
+
+/ {
+	soc {
+
+		uart@f0004000 {
+			interrupts = <108 1>;
```
Comment (galak) [inline boards/arc/emsdp/emsdp_em4.dts]: Is this really a board level change?  Feels like this should possible be in a soc dtsi file?

File: soc/arc/snps_emsdp/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_compile_options(-mcpu=${GCC_M_CPU} -mno-sdata)
+
+if(CONFIG_SOC_EMSDP_EM4)
```
Comment (galak) [inline soc/arc/snps_emsdp/CMakeLists.txt]: Feels like this should get refactored between the other arc SoC's, but no issue.

File: soc/arc/snps_emsdp/soc.c [hunk 9]
```diff
@@ -0,0 +1,12 @@
+
+/*
+ * Copyright (c) 2019 Synopsys, Inc. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Comment (galak) [inline soc/arc/snps_emsdp/soc.c]: if file is empty, why have it?

**Review by galak — DISMISSED** [review]
Various comments and some changes.

File: boards/arc/emsdp/emsdp_em4.dts (hunk 7 above)
Comment (IRISZZW) [reply boards/arc/emsdp/emsdp_em4.dts]: emsdp board is similar to em_starterkit board, it is FPGA based board which supports different core configurations, such as EM4,
EM5D, EM6, EM7D, EM9D, EM9D+ESP, EM11D. in different core configurations the interrupts number may be different. so we need to point it out in core configuration related dts file.

File: soc/arc/snps_emsdp/CMakeLists.txt (hunk 8 above)
Comment (IRISZZW) [reply soc/arc/snps_emsdp/CMakeLists.txt]: different core configuration has different compile options.

File: soc/arc/snps_emsdp/soc.c (hunk 9 above)
Comment (IRISZZW) [reply soc/arc/snps_emsdp/soc.c]: to make code organization structure uniformed with other arc core board.

Comment (galak) [reply soc/arc/snps_emsdp/soc.c]: we should remove the file.

File: boards/arc/emsdp/emsdp_em4.dts (hunk 7 above)
Comment (galak) [reply boards/arc/emsdp/emsdp_em4.dts]: Since these are different core/SoC I think this should move into dts/arc/em{5d,6,7,9d,9d+esp,11d}.dtsi

Comment (vonhust) [reply boards/arc/emsdp/emsdp_em4.dts]: Currently, most of differences of  different core are board-specific. Not easy to move into dts/arc/

We know there are lots of duplications,  plan to do a cleanup with other arc boards together in future.

File: soc/arc/snps_emsdp/soc.c (hunk 9 above)
Comment (IRISZZW) [reply soc/arc/snps_emsdp/soc.c]: removed

**Comment by vonhust** [comment]
@galak @ruuddw @carlescufi  ping for merge if no other issues.

**Comment by carlescufi** [comment]
@galak please merge at your convenience

**Comment by vonhust** [comment]
ping for merge, @carlescufi @galak @ioannisg @nashif

**Pull request merged**